### ISSUE TITLE
mut_mut complains when borrowing from a slice of mutable references

### ISSUE THREAD

**Issue by cmrschwarz** [rust-lang/rust-clippy#13021]
### Summary

There are many valid reasons to work on a mutable slice of mutable references `&mut [&mut Foo]`.

For example, if the `Foo`s are large and you temporarily want to work on them in a sorted order,
you might reasonably want to sort mutable references to them instead.
In other cases the `Foo`s might not be movable at all, but you want to process a shrinking subset of them.

When you now want to borrow a single element from this slice,
you end up with the type `&mut &mut Foo`,
because you can't move / clone the mutable reference out of the slice.

Clippy complains about this, which it shouldn't imo.

### Lint Name

mut_mut

### Reproducer

I tried this code:

```rust
#![warn(clippy::pedantic)]
fn process_foos(array_of_foos: &mut [&mut Foo]) {
    let entry = &mut array_of_foos[42];
    // modify entry...
}
```

I saw this happen:

```
warning: this expression mutably borrows a mutable reference. Consider reborrowing
  --> src/main.rs:13:17
   |
13 |     let entry = &mut array_of_foos[42];
   |                 ^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#mut_mut
note: the lint level is defined here
  --> src/main.rs:9:8
   |
9  | #[warn(clippy::pedantic)]
   |        ^^^^^^^^^^^^^^^^
   = note: `#[warn(clippy::mut_mut)]` implied by `#[warn(clippy::pedantic)]`
```

I expected to see this happen:
No warnings should be generated.

### Version

```text
rustc 1.78.0 (9b00956e5 2024-04-29)
binary: rustc
commit-hash: 9b00956e56009bab2aa15d7bff10916599e3d6d6
commit-date: 2024-04-29
host: x86_64-unknown-linux-gnu
release: 1.78.0
LLVM version: 18.1.2
```

**Comment by ada4a** [comment]
but reborrowing _would_ work here, no?
```rs
let entry = &mut *array_of_foos[42];
```

**Comment by cmrschwarz** [comment]
Looking at the code that caused me to report this again, you're indeed correct. 
Sorry for the incorrect report!

I think at the time I only tried  `let entry = array_of_foos[42];`, which doesn't work,
because the inferred lifetime of the borrow would be too long.

In my defense, for somebody who was newer to Rust at the time, this is pretty unintuitive. Especially since both expressions yield the same "visual type" (`&mut Foo`), and if you are a bit fuzzy on terminology it is easy to think of what you are trying
to do (`&mut &mut Foo`) as "reborrowing", leading to confusion about the lint message.

I think if this lint included a fix suggestion it would go a long way in helping new users, such as my past self.

Feel free to close this though!

**Comment by ada4a** [comment]
Didn't expect to get such a quick response and for you to remember the context behind this! Thank you for a detailed answer:) 

I very much agree that having a suggestion would help – I'll try to go implement it.

@rustbot claim
